src/util/AsyncStream.h

```diff
@@ -78,6 +82,56 @@ cppcoro::generator<typename Range::value_type> runStreamAsync(
                << "ms" << std::endl;
   });
 }
+
+/**
+ * This is iterator based runStreamAsync template to facilitate refactoring from
+ * cppcoro::generator<ValueType> to ad_utility::InputRangeTypeErased<ValueType>
+ */
+template <typename Range, bool logTime = (LOGLEVEL >= TIMING)>
+ad_utility::InputRangeTypeErased<typename Range::value_type> runStreamAsyncV2(
+    Range range, size_t bufferLimit) {
+  using value_type = typename Range::value_type;
+
+  struct Generator : public ad_utility::InputRangeFromGet<value_type> {
+    Generator(Range range, const size_t bufferLimit)
+        : queue{std::make_shared<ThreadSafeQueue<value_type>>(bufferLimit)},
+          exception{std::make_shared<std::exception_ptr>(nullptr)},
+          thread{std::shared_ptr<std::thread>(
+              new std::thread(&Generator::thread_function, std::move(range),
+                              queue, exception),
```

Comment:
1. Use `ad_utility::JThread` (automatic join) and don't use plain `new` / `delete` which makes problems, `shared_ptr` automatically does that for you

2. Why do you use `shared_ptr` at all here and not a plain value? (same for the `ThreadSafeQueue`? 
 I assume the problem is, that this generator doesn't  have correct move semantics and you thus run into trouble, when placing it into an `TypeErasedInputRange` , but there we should have some better interfaces (1. unique_ptr instead of shared_ptr at least, then we automatically get move only etc. Also I assume that some of your problems go away if you drop the manual `delete` call.

Reply:
the reason is ThreadSafeQueue, it has copy and move disabled, what to do with this?

Reply:
@joka921 please take a look at [thiscommit](https://github.com/ad-freiburg/qlever/pull/2151/commits/e750e0be56fc5a25411ad5d3b74a4284731769a1) std::muted is not moveable so, I splitted the generator for core and moveable struct, how do you like it?

Reply:
ThreadSafeQueue refactored to use the same mechanism in this [commit](https://github.com/ad-freiburg/qlever/pull/2151/commits/d1f12b45804f2d055d7d46606f034a82f09b963d)